plugin.video.viaplay/resources/lib/kodihelper.py

```diff
@@ -0,0 +1,228 @@
+        elif country_id == '1':
+            country_code = 'dk'
+        elif country_id == '2':
+            country_code = 'no'
+        else:
+            country_code = 'fi'
+
+        return country_code
+
+    def get_sub_lang(self):
+        sub_lang_id = self.get_setting('sub_lang')
+        if sub_lang_id == '0':
+            sub_lang = 'sv'
+        elif sub_lang_id == '1':
+            sub_lang = 'da'
+        elif sub_lang_id == '2':
+            sub_lang = 'no'
+        else:
+            sub_lang = 'fi'
+
+        return sub_lang
+
+    def dialog(self, dialog_type, heading, message=None, options=None, nolabel=None, yeslabel=None):
+        dialog = xbmcgui.Dialog()
+        if dialog_type == 'ok':
+            dialog.ok(heading, message)
+        elif dialog_type == 'yesno':
+            return dialog.yesno(heading, message, nolabel=nolabel, yeslabel=yeslabel)
+        elif dialog_type == 'select':
+            ret = dialog.select(heading, options)
+            if ret > -1:
+                return ret
+            else:
+                return None
+
+    def authorize(self):
+        try:
+            self.vp.validate_session()
+            return True
+        except self.vp.ViaplayError as error:
+            if not error.value == 'PersistentLoginError' or error.value == 'MissingSessionCookieError':
+                raise
+            else:
+                return self.device_registration()
+
+    def log_out(self):
+        confirm = self.dialog('yesno', self.language(30042), self.language(30043))
+        if confirm:
+            self.vp.log_out()
+            # send Kodi back to home screen
+            xbmc.executebuiltin('XBMC.Container.Update(path, replace)')
+            xbmc.executebuiltin('XBMC.ActivateWindow(Home)')
+
+    def device_registration(self):
+        """Presents a dialog with information on how to activate the device.
+        Attempts to authorize the device using the interval returned by the activation data."""
+        activation_data = self.vp.get_activation_data()
+        message = self.language(30039).format(activation_data['verificationUrl'], activation_data['userCode'])
+        dialog = xbmcgui.DialogProgress()
+        xbmc.sleep(200)  # small delay to prevent DialogProgress from hanging
+        dialog.create(self.language(30040), message)
+        secs = 0
+        expires = activation_data['expires']
+
+        while secs < expires:
+            try:
+                self.vp.authorize_device(activation_data)
+                dialog.close()
+                return True
+            except self.vp.ViaplayError as error:
+                # raise all non-pending authorization errors
+                if not error.value == 'DeviceAuthorizationPendingError':
+                    raise
+            secs += activation_data['interval']
+            percent = int(100 * float(secs) / float(expires))
+            dialog.update(percent, message)
+            xbmc.sleep(activation_data['interval'] * 1000)
```

Comment:
Same here, `xbmc.Monitor().waitForAbort(seconds)` is preferable

Reply:
@enen92 I'm unsure on where I should implement the `waitforAbort` method. This is my loop with `abortRequested`:

```python
        while not xbmc.Monitor().abortRequested() and secs < expires:
            try:
                self.vp.authorize_device(activation_data)
                dialog.close()
                return True
            except self.vp.ViaplayError as error:
                # raise all non-pending authorization errors
                if not error.value == 'DeviceAuthorizationPendingError':
                    raise
            secs += activation_data['interval']
            percent = int(100 * float(secs) / float(expires))
            dialog.update(percent, message)
            xbmc.sleep(activation_data['interval'] * 1000)
            if dialog.iscanceled():
                dialog.close()
                return False

        dialog.close()
        return False
```

Reply:
Instead of xbmc.sleep basically. The sleep method should only be used for really small intervals

Reply:
@enen92 Ah, got it!